File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
    let dummy = new ListNode(0, head) // dummy -> 1-> 2 ...
    let left = dummy
    let right = head
    while (n > 0 && right !== null) {
        right = right.next
        n -= 1
    }

    while (right !== null) {
        left = left.next
        right = right.next
    }

    left.next = left.next.next
    return dummy.next
};
// 1 2 3 4 5
// 0 1 2 3 4 5 null
// L     R
//   L     R
//     L     R
//       L      R

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        # given head of linked list and an integer n
        # return head of linked list after removing nth node FROM THE END of the list
        # to remove nodes we want to do head.next = head.next.next
        # so as long as we find the node right before the nth node from the end
        # we can remove easily.

        result = ListNode(0, head)
        dummy = result

        for _ in range(n):
            head = head.next
        
        while head:
            head = head.next
            dummy = dummy.next
        
        dummy.next = dummy.next.next

        return result.next
